### PR TITLE
Initial attempt at a test framework

### PR DESCRIPTION
**Pull request by isovector**
This PR adds an hspec test suite, which uses [hspec-discover](https://hspec.github.io/hspec-discover.html) to automatically find tests to run.

Run it via:

```
stack test concat-plugin:tests
```

It introduces a free syntax category, to which I compile simple lambdas and inspect the resulting structure. Most of these tests fail, although I'm not sure if my understanding or the plugin is to blame. I suspect the former :)

I'm particularly perplexed by the final test. AFAICT it is a copy of the misc-example `double` circuit (which succeeds), while this fails.

### CODE REVIEW

File: plugin/test/Types.hs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+import qualified ConCat.Category as CC
+import           Data.Typeable
+import           Unsafe.Coerce (unsafeCoerce)
+
+
+----------------------------------------------------------------------------
+-- | The free category construction over all of the relevant concat classes.
+data FreeSyn a b where
+  CId    :: FreeSyn a b
```
Comment (conal) [inline plugin/test/Types.hs]: Do you want `CId    :: FreeSyn a a` instead?

File: plugin/test/Types.hs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  show CNeg        = "neg"
+  show CInl        = "inl"
+  show CInr        = "inr"
+  show (CCOr a b)  = "(or " ++ show a ++ " " ++ show b ++ ")"
+  show (CConst a)  = "(const " ++ show a ++ ")"
+
+
+instance Eq (FreeSyn a b) where
+  CId == CId = True
+  CComp a b == CComp c d =
```
Comment (conal) [inline plugin/test/Types.hs]: You could probably define `Ok FreeSyn` to eliminate the need for `unsafeCoerce`.

File: .gitignore [hunk 3]
```diff
@@ -7,3 +7,5 @@
 *.o
 *.pdf
 o[1-9]*
+tags
```
Comment (conal) [inline .gitignore]: I put always-ignore patterns like "tags" and "TAGS" into my personal global git ignore file, so I don't have to add it to per-repo ignore files. I don't know how common this practice is.

File: plugin/test/SimpleSpec.hs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      `shouldBe`
+        CInl
+
+  -- fails: "Oops: toCcc' called"
+  it "inr" $
+    toCcc Right
+      `shouldBe`
+        CInr
+
+  -- -- fails to compile: "panic! lam Case of boxer: bare unboxed var"
```
Comment (conal) [inline plugin/test/SimpleSpec.hs]: You probably need another import:

``` haskell
import ConCat.Rebox () -- necessary for reboxing rules to fire
```

This import exposes rewrite rules that perform reboxing and some other translations.

**Comment by conal** [comment]
Some general comments:

These tests don't seem to tickle the kinds of bugs I encounter, most of which are either a compile-time panic or a run-time exception ("Oops: toCcc' called"). The tests are also sensitive to optimizations GHC legitimately makes. I wonder how we could accommodate both failures and something closer to denotational equality tests (equality of meanings rather than representations).

If we do want a GADT-based syntactic category, maybe we merge it into the syntactic category in `ConCat.Syntactic` (described in the paper), i.e., have a GADT representation and add a pretty-printer.

File: plugin/test/SimpleSpec.hs (hunk 4 above)
Comment (isovector) [reply plugin/test/SimpleSpec.hs]: This is particularly weird in terms of usability. Could it be re-exported from somewhere else, or injected into the module via the plugin?

**Comment by isovector** [comment]
Just to be explicit, I think the value of tests like these (although not necessarily *these*) is three-fold:

* Preventing regressions
* Having a framework in place helps ensure new features are adequately tested
* Serving as simple documentation for onboarding new contributors

When you say "sensitive to optimizations GHC makes", are those optimizations at least deterministic? If so, it strikes me as the kind of thing it might be valuable to unit test. 

I know of no ways to test for compile-time panics (unless we can maybe CCC the plugin itself, and inspect that for panic-reachability?), but run-time exceptions seem to me like they're best caught with a test CCCing a lot of diverse examples. For example, if all of `base` CCCs correctly (and there is sufficient trust in composition), that's a pretty good sign.

Testing on a denotational level I'd suspect to be a lot harder; lambda terms are notoriously difficult to compare for equality. Extensionally, we could re-interpret the syntax as a lambda, and then quick-check the two implementations for equality, but I doubt we could do much better than this approach intentionally.

Unless you disagree (or have a specific proposal for a better heading), I'd like to push forward in this direction -- if for no other reason than to have an easily-runnable, easily-inspectable suite of concat's capabilities.

**Comment by isovector** [comment]
After reading through the paper and `Plugin.hs` a few more times, I think I know what's going on here with my perplexing tests:

1) The `exl` and `exr` tests are conceptually correct, but the `lam Case of product` rule is firing, and there is no special logic to emit `exl` directly for `fst` and v.v.
2) There is no corresponding `lam Case of coproduct` which is why eg. `tail` seems to refuse to CCC.
3) No logic seems to currently exist to transform `Left -> inl`, likewise for `Right -> inr`.

If these suspicions are true, (2) and (3) suggest there exist category classes which are currently untargetable. Is this so? If yes, does there exist documentation anywhere about what is done and what isn't?

**Comment by conal** [comment]
Sorry for the delay in replying. I'm on a driving trip and have limited time online.

**Comment by conal** [comment]
> The `exl` and `exr` tests are conceptually correct, but the `lam Case of product` rule is firing, and there is no special logic to emit `exl` directly for `fst` and v.v.

Right. I think I used to have some such short-cuts, but I removed them. I prefer to handle these situations via rewrite rules that are true for all categories, cartesian categories, etc., for maximum benefit. I'll see what I can do.

> 2) There is no corresponding `lam Case of coproduct` which is why eg. `tail` seems to refuse to CCC.
> 3) No logic seems to currently exist to transform `Left -> inl`, likewise for `Right -> inr`.
> 
> If these suspicions are true, (2) and (3) suggest there exist category classes which are currently untargetable. Is this so? If yes, does there exist documentation anywhere about what is done and what isn't?

Your suspicions about coproducts are true. I have translation to coproducts in an earlier version of the plugin but haven't yet added it to the current version. There's not much to it (though it does involve `DistribCat` as well as `CoproductCat`). I don't think there are other category classes not yet targeted by the plugin.

**Pull request closed without merging**